mne/preprocessing/_csd.py

```diff
@@ -0,0 +1,169 @@
+                                   stiffness=4, n_legendre_terms=50,
+                                   sphere=(0., 0., 0., 0.095),
+                                   copy=True):
+    """Current Source Density (CSD) transformation
+
+    Transformation based on spherical spline surface Laplacian.
+
+    .. note:: This function applies an average reference to the data.
+              Do not transform CSD data to source space.
+
+    Parameters
+    ----------
+    inst : instance of Raw, Epochs or Evoked
+        The data to be transformed.
+    picks : np.ndarray | list, shape (n_channels,) | None
+        The picks to be used. Defaults to None.
+    lambda2 : float
+        Regularization parameter, produces smoothnes. Defaults to 1e-5.
+    stiffnes : float
+        stiffnes of the spline. Also referred to as `m`
+        (if g_matrix or h_matrix is provided this will be ignored)
+    n_legendre_terms : int
+        number of Legendre terms to evaluate (if g_matrix or h_matrix
+        is provided this will be ignored)
+    sphere : tuple, (float, float, float, float)
+        The sphere, head-model of the form (x, y, z, r) where x, y, z
+        is the center of the sphere and r is the radius
+    copy : bool
+        Whether to overwrite instance data or create a copy.
+
+    Returns
+    -------
+    inst_csd : instance of Epochs or Evoked
+        The transformed data. Output type will match input type.
+    """
+
+    _validate_type(inst, (BaseEpochs, BaseRaw, Evoked), 'inst')
+
+    if 'csd' in inst.info['comps']:
+        raise ValueError('CSD already applied, should not be reapplied')
+
+    inst = inst.copy() if copy else inst
+
+    if picks is None:
+        picks = pick_types(inst.info, meg=False, eeg=True, exclude='bads')
+    _validate_type(picks, (list, np.ndarray), 'picks')
+
+    if len(picks) == 0:
+        raise ValueError('No EEG channels found.')
+
+    if lambda2 is None:
+        lambda2 = 1e-5
+
+    _validate_type(lambda2, (float, int), 'lambda2')
+    if 0 > lambda2 or lambda2 > 1:
+        raise ValueError('lambda2 must be between 0 and 1, got %s' % lambda2)
+
+    _validate_type(stiffness, (float, int), 'stiffness')
+    if stiffness < 0:
+        raise ValueError('stiffness must be non-negative got %s' % stiffness)
+
+    _validate_type(n_legendre_terms, (int), 'n_legendre_terms')
+    if n_legendre_terms < 1:
+        raise ValueError('n_legendre_terms must be greater than 0, '
+                         'got %s' % n_legendre_terms)
+
+    _validate_type(sphere, tuple, 'sphere')
+    x, y, z, radius = sphere
+    _validate_type(x, float, 'x')
+    _validate_type(y, float, 'y')
+    _validate_type(z, float, 'z')
+    _validate_type(radius, float, 'radius')
+    if radius <= 0:
+        raise ValueError('sphere radius must be greater than 0, '
+                         'got %s' % radius)
+
+    _validate_type(copy, (bool), 'copy')
+
+    pos = np.array([inst.info['dig'][pick]['r'] for pick in picks])
```

Comment:
we should really use `info['chs'][pick]['loc'][:3]` here, as `dig` may not only contain EEG (likely contains fiducials), and we should also check that no channel is all `0.` and all are `np.isfinite`